csrc/runtime/executor.cpp

```diff
@@ -410,151 +234,56 @@ void KernelExecutor::compile(
   } else if (has_cp_async_bulk) {
     // TMA operations require 32-bit indexing.
     compile_params.index_type = PrimDataType::Int32;
+  } else {
+    compile_params.index_type = arg_index_type;
   }
 
-  c10::DeviceGuard dg(options_.device);
+  c10::DeviceGuard dg(device);
 
-  NVF_ERROR(
-      options_.device.is_cuda(), "Provided device to CUDA fuser is the CPU.");
-  auto properties = at::cuda::getDeviceProperties(options_.device.index());
+  NVF_ERROR(device.is_cuda(), "Provided device to CUDA fuser is the CPU.");
+  auto properties = at::cuda::getDeviceProperties(device.index());
   // TODO: These properties should be set as part of the constructor so that it
   // can be const
   device_smem_limit_ = static_cast<int64_t>(properties->sharedMemPerBlockOptin);
   warp_size_ = properties->warpSize;
 
-  lowered_ = std::make_unique<GpuLower>(fusion, compile_params);
-  for (const auto& hook : lowering_hooks_) {
-    hook(lowered_.get());
-  }
-  lowered_->run();
-
-  kir::Kernel* kernel = lowered_->kernel();
-
-  for (const auto& hook : post_lowering_hooks_) {
-    hook(kernel);
-  }
-
-  setUsedTVs();
-
-  if (isDebugDumpEnabled(DebugDumpOption::KernelIr)) {
-    kernel->print();
-  }
-
-  if (isDebugDumpEnabled(DebugDumpOption::BankConflictInfo)) {
-    auto bank_conflict_info = getBankConflictInfo(kernel);
-    if (bank_conflict_info.empty()) {
-      debug() << "===== No bank confliction =====" << std::endl;
-    } else {
-      debug() << "======= Bank confliction =======" << std::endl;
-      for (auto info : bank_conflict_info) {
-        debug() << "Expr: " << info.first->toString() << std::endl;
-        auto conflict = info.second;
-        if (conflict.first > 1) {
-          debug() << "input conflict: " << conflict.first << " way, ";
-        }
-        if (conflict.second > 1) {
-          debug() << "output conflict: " << conflict.second << " way";
-        }
-        debug() << std::endl;
-      }
-      debug() << "================================" << std::endl;
-    }
-  }
+  // Lowered is needed to compute launch parameters as it uses the CA map. We
+  // could modify that, but simply generating that part first.
+  compiled_kernel_ = std::make_unique<CompiledKernel>(
```

Comment:
I'm trying to understand the supposed responsibility between KernelExecutor::compile and CompiledKernel::compile. With this PR, KernelExecutor::compile appears to become a thin wrapper of CompiledKernel::compile that merely does profiling and overrides some compilation/launch parameters. Do you plan to get rid of KernelExecutor::compile so FusionKernelRuntime doesn't need to create a KernelExecutor to compile?

Reply:
I was wondering the same thing, and I think ideally yes. If I recall correctly the challenge comes down to requiring runtime information to compile. It makes it a bit awkward in my opinion as it'd be nice to have a clearer boundary between compilation and runtime, but we need runtime information for compilation. What do you think?

Reply:
Sounds reasonable, how should we handle needed recompilation under dynamic workloads when there's a heuristic match?

https://github.com/NVIDIA/Fuser/pull/3468/files#diff-4e98fb6577fe11f6854ee230d2ed5952b441fa277061180b00fa1265cf98173eR1527-R1552

Should this also be represented in HostIR? Should we have a "maybe recompile" node?

Reply:
TIL: I didn't know about KernelExecutor::recompileKernel. AFAICT, it recompiles CUDA with a new block size and/or a new max number of registers, but doesn't re-schedule, re-lower or re-codegen. 

I'd keep this "mini-caching" in CompiledKernel. We could represent this recompilation in host IR, but I don't see a practical benefit -- having a "maybe recompile" node doesn't enable more host IR optimization or make the result host program faster. What do you think?

Reply:
I think your plan is good. I think this PR would help as it's a step towards isolating compilation from execution. I think 1 is doable with your plan. I think you will still need a maybe recompile in the execution somewhere if launch parameters change. Right now it's being called in FusionExecutor::runFusion: https://github.com/NVIDIA/Fuser/blob/main/csrc/runtime/executor.cpp#L1269

The reason we do this is the same reason why someone might use __launch_bounds__ on a cuda kernel. The first time we get a problem we might compile with one register limit in mind based on the number of threads we first compute are needed. If that number goes up because it's input dependent then we may need to recompile otherwise the kernel won't fit on the SM.